### ISSUE TITLE
Cell probabilities remain close to one

### ISSUE THREAD

**Issue by leprohonmalo** [broadinstitute/CellBender#220]
Hi,

Thank you for developing this tool ! I am trying CellBender (using latest docker image) on several human samples containing mostly T cells. We loaded 20 000 cells on the chip aiming to recover 7000-10 000.

![newplot](https://github.com/broadinstitute/CellBender/assets/45560569/bbf234b7-0726-4c0f-ba80-3b49982604e2)

For many samples (3/4) I get similar results where all droplets have a cell probabilities close to one.

![learning](https://github.com/broadinstitute/CellBender/assets/45560569/ca5095fb-e0f6-42a8-9cdc-2f7c3b23c476)

![cell-prob](https://github.com/broadinstitute/CellBender/assets/45560569/1301b34e-2686-4f20-9fe4-b54ccd10c525)

![pca](https://github.com/broadinstitute/CellBender/assets/45560569/b33e1488-b6ba-4703-bd22-2ab2f80bf63e)

I runned my sample with the following code : 

```
cellbender remove-background \
	--input raw_feature_bc_matrix.h5 \
	--cellbender_default.h5 \
	--expected-cells 10000 \
	--total-droplets-included 35000 \
	--low-count-threshold 15 \
	--empty-drop-training-fraction 0.5 \
	--z-dim 100 \
	--z-layer 500 \
	--fpr 0.01 \
	--epochs 150 \
	--cuda
```
I saw similar issues suggesting to adjust parameters such as :  
```--expected-cells``` (tested 3000; 5000; 10 000)
```--total-droplets-included``` (tested from 10 000 to 40 000)
```--low-count-threshold``` (tested 5; 15)
```--empty-drop-training-fraction``` (tested 0.3; 0.5; 0.7)
```--z-dim``` (tested 50; 100)
```--z-layers``` (tested 250; 500)

However I could not manage to make it work. 
What could I do to improve the results ? Is there any issue that I am not seeing  (for this specific sample the barcode rank plot maybe has an unusual shape and I recover a low amount of cells but that is not applicable to all of my samples) ? 
Thank you in advance for your help. Tell me if I can add any useful information.

**Comment by aodainic7** [comment]
I am not a developer, but based on the graph you show, it seems like the algorithm does has not enough droplets to reach the empty plateau, it stops at 25k total droplets. Try going higher with the `--total-droplets-included`, like 70k.

**Comment by sjfleming** [comment]
Hi @leprohonmalo ,
Sorry you've been struggling with this: I am aiming to make the user experience better in v0.3.0...
It seems like you have already tried quite a bit of parameter tweaking.

So, from looking at this sample's UMI curve (the top plot), I have kind of a hard time knowing exactly what's going on with the experiment.  There are two possible scenarios:

1. There are like 3k cells, and then there is an empty droplet plateau that goes out until approximately droplet 10k.  The empty droplets have something like 500 UMI counts.
2. There are like 10k cells, and they fall into two very different groups: those with 5k+ UMI counts, and those with about 500 UMI counts.  Then the empty droplet plateau is after droplet 10k.  The empty droplets have something like 10 UMI counts.

Evidence for scenario (1):
- 3k cells would be in the right ballpark
- 500 counts in empty droplets would represent a very high amount of ambient RNA, but I have seen samples like that before

Evidence against scenario (1):
- If the empty droplets are between droplet 3k and 10k on the UMI curve, then why are there so few empties?  For a 10x experiment, I expect to see many tens of thousands of empty droplets.  Was there some reason that you would expect to see fewer empty droplets than usual in this experiment?  Did you not use all the 10x GEM barcoded beads?
- 500 counts is a very high amount of ambient RNA, and if these are T cells captured from solution, then I would not expect very high ambient RNA unless something else is going on.  PBMC datasets typically have very low amounts of ambient RNA.

Evidence for scenario (2):
- 10 UMI counts in empty droplets is a lot closer to what we see in the public 10x Genomics PBMC datasets that use v3 chemistry
- There would be tens of thousands of empty droplets, which is what I expect to see.

Evidence against scenario (2):
- Do you expect to see two populations of cells, maybe 3k of one type and 7k of another type, where the first cell type has 5k+ UMI counts and the second cell type has only 500 UMI counts?  Is this all identical T cells, or are there other things in here like granulocytes, which can have very low UMI counts like that?
- The amount of ambient RNA would be very low, suggesting this was a very clean experiment.  Is that consistent with your expectations?


Are you measuring just RNA here, or are there antibody features or ATAC features as well?

**Comment by sjfleming** [comment]
Hi @leprohonmalo , thanks for your comments, glad to hear those kind things.

I think I tend to agree with your assessment after hearing your answers.  If this was a "normal" 10x run with PBMCs, then I would expect to see an empty droplet plateau going out to around 70k droplets or so (which is what you see), and from blood with PBMC isolation, I'd also expect the experiment to be pretty clean: I would not expect to see 500 UMI counts in empty droplets (which is what I usually see from very difficult snRNA-seq preps from tough human tissues like heart or blood vessel walls).

(Another thing you might be able to check is this: in the raw data, if you look at droplets from the 500 UMI count plateau, do they express new genes that are NOT present in the droplets with 2000+ counts?  If the 500 count plateau was really empty droplets, you would not expect to see new genes that didn't come from the cells in the experiment.)

So if scenario 2 is the more likely, then I would recommend trying to run cellbender as follows:

```
cellbender remove-background \
        --cuda \
	--input raw_feature_bc_matrix.h5 \
	--output cellbender_default.h5 \
	--expected-cells <number before the true empty droplet plateau> \
	--total-droplets-included <number after the empty droplet plateau> \
	--low-count-threshold 5 \
	--fpr 0.01
```
where `<number before the true empty droplet plateau>` would be like 10000 for the CellRanger UMI curve you show at the top, and where `<number after the empty droplet plateau>` would be like 35000.  However, for that sample you show the cellbender output for, you would probably need larger values for both of those.  It looks like there really are 25k+ cells, and the true empty droplets are not included in the analysis, because I do not see the UMI curve fall off to the ambient RNA level of around 10 or 20 or 30 counts.

The `--low-count-threshold` value of 5 will be important for your experiments, since the default value is 15 is too large for your experiments where the true empty droplets have only 10 counts or so.

**Comment by leprohonmalo** [comment]
Hi @sjfleming,

I looked at the 500 UMI population and it is definitly not a second population of cells as it displays mostly expression of mitochondrial genes (around 50 %). So I was wrong and it seems that scenario 1 is much more likely. What would you advise in this case ?

**Comment by sjfleming** [comment]
Well, I guess I would not be so sure that the high mitochondrial reads is evidence for scenario 1.  In fact, if 50% mitochondrial reads is higher than the fraction in the really good cells, then I would suspect that the 500 UMI plateau is NOT empty droplets... precisely because there's something _different_ going on in those 500 UMI count droplets than in the cells.

It's kind of like this sort of scenario
https://github.com/broadinstitute/CellBender/issues/121#issuecomment-1379425406

It's hard to say what those droplets actually represent.  In my mind, they could be either dying cells (which sometimes have a high mitochondrial read fraction) or they could be some kind of debris-containing droplets that for some reason contain a lot of mitochondria.  But I think, for the purposes of CellBender's model, they are not empty.  So I would think scenario 2 still applies.

**Comment by leprohonmalo** [comment]
Thanks for the explanation.
I tested a few more runs with an increase in the parameters values you suggested and I am starting to get somewhere ! (also thanks for your suggestion @aodainic7. At first I thought it was too much and that 40000 already looked like in the empty droplet plateau from cellranger plot, but it proved quite relevant) 
Parameters : 
```
cellbender remove-background \
        --input raw_matrix/CART_P88_raw_feature_bc_matrix.h5 \
        --output results/CART_P88_cellbender_50k_25k_3.h5 \
        --expected-cells 25000 \
        --total-droplets-included 50000 \
        --low-count-threshold 3 \
        --fpr 0.01 \
        --epochs 150 \
        --cuda
```
![image](https://github.com/broadinstitute/CellBender/assets/45560569/56be594d-6bed-4024-b334-3b228f78df0d)

![image](https://github.com/broadinstitute/CellBender/assets/45560569/02d57bb4-9a38-46da-ae20-75ab7f3235c8)

However the ELBO plot is not perfect, do you think that I can improve it with parameters like learning-rate ? Is it fine to go on with it ? I would say that previous epochs still converged.

**Comment by sjfleming** [comment]
So I can tell you what's happening here... and it's something I'm currently struggling to fix.

The turnaround in the learning curve is due to those initial few droplets with the highest counts getting called "empty" by cellbender.  This is something that ideally should NOT happen.  And I usually see a kind of dip in the learning curve when it does happen.  I'm working on incorporating some more regularization to make sure it doesn't happen in future versions.

(The output -- apart from those few high count droplets -- should be fine though.  ...apart from the handful of high-count droplets that are actually being called "empty": those droplets will have zero counts in the output, which is probably not desirable... though those droplets probably represent doublets anyway... but still it would be better if cellbender would keep them.)

For now, the options would be (1) to reduce the learning rate, maybe `--learning-rate 2e-5`, or (2) to reduce the number of epochs of training and just stop earlier, maybe `--epochs 100`.  (Fewer epochs won't necessarily be equivalent to stopping early on the previous learning curve though... we use a learning rate scheduler, so changing the number of epochs changes the learning rate schedule, and it can be a little less predictable than you'd think.)

**Comment by leprohonmalo** [comment]
Hi,
I tried what you suggested and it worked like a charm ! 

[CART_P88_cellbender_50k_25k_5_2e5.pdf](https://github.com/broadinstitute/CellBender/files/11723837/CART_P88_cellbender_50k_25k_5_2e5.pdf)

Thank you very much for your help and your explanations. I hope I will be able to run the rest of  my samples now. I think I'm good with closing the issue.
